From f83b5d478d8b0921efbdfd11fc19cd1122128449 Mon Sep 17 00:00:00 2001
From: Abdullah Yousufi <abdullah.yousufi@cloudera.com>
Date: Thu, 14 Jul 2016 14:47:56 -0500
Subject: [PATCH 0672/1164] CDH-41578: HIVE-14074: RELOAD FUNCTION should
 update dropped functions (Abdullah Yousufi,
 reviewed by Sergey Shelukhin, Sergio Pena)

Change-Id: If38620b53f09aab3a46a5bd3015e1ed5b5122027
---
 .../org/apache/hadoop/hive/ql/metadata/Hive.java   |   18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
index 5007c06..e3e68d3 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
@@ -212,18 +212,30 @@ private void registerAllFunctionsOnce() throws HiveException {
   }
 
   public void reloadFunctions() throws HiveException {
+    HashSet<String> registryFunctions = new HashSet<String>(
+        FunctionRegistry.getFunctionNames(".+\\..+"));
     for (Function function : getAllFunctions()) {
       String functionName = function.getFunctionName();
       try {
         LOG.info("Registering function " + functionName + " " + function.getClassName());
-        FunctionRegistry.registerPermanentFunction(
-                FunctionUtils.qualifyFunctionName(functionName, function.getDbName()), function.getClassName(),
-                false, FunctionTask.toFunctionResource(function.getResourceUris()));
+        String qualFunc = FunctionUtils.qualifyFunctionName(functionName, function.getDbName());
+        FunctionRegistry.registerPermanentFunction(qualFunc, function.getClassName(), false,
+                    FunctionTask.toFunctionResource(function.getResourceUris()));
+        registryFunctions.remove(qualFunc);
       } catch (Exception e) {
         LOG.warn("Failed to register persistent function " +
                 functionName + ":" + function.getClassName() + ". Ignore and continue.");
       }
     }
+    // unregister functions from local system registry that are not in getAllFunctions()
+    for (String functionName : registryFunctions) {
+      try {
+        FunctionRegistry.unregisterPermanentFunction(functionName);
+      } catch (Exception e) {
+        LOG.warn("Failed to unregister persistent function " +
+            functionName + "on reload. Ignore and continue.");
+      }
+    }
   }
 
   public static Hive get(Configuration c, Class<?> clazz) throws HiveException {
-- 
1.7.9.5

